### PR TITLE
[ci-app] Add Error Handling for `beforeEach`/`afterEach` Hooks in Testing Frameworks

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?
* Handle errors in hooks for `jest`, `mocha` and `cucumber`. 

### Motivation
If a test fails on its `beforeEach` or `afterEach`, we still want to have visibility over that error. This PR addresses just that.

### Plugin Checklist

- [x] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/src/util.js [hunk 1]
```diff
@@ -15,7 +15,9 @@ function isError (value) {
     return true
   }
   if (value && value.constructor) {
-    return value.constructor.name === 'JestAssertionError' || value.constructor.name === 'Error'
+    return value.constructor.name === 'JestAssertionError' ||
+      value.constructor.name === 'Error' ||
+      value.constructor.name === 'ErrorWithStack'
```
Comment (juan-fernandez) [inline packages/dd-trace/src/util.js:20]: again, something weird with errors and `jest`: https://github.com/facebook/jest/blob/3b253f83b7500f2d5b627f4ab5f6971e7ef8349c/packages/jest-util/src/ErrorWithStack.ts#L8

File: packages/datadog-plugin-cucumber/test/index.spec.js [hunk 2]
```diff
@@ -180,6 +195,10 @@ describe('Plugin', () => {
                 const parentCucumberStep = testTrace.find(span => span.meta['cucumber.step'] === 'integration')
                 expect(httpSpan.parent_id.toString()).to.equal(parentCucumberStep.span_id.toString())
               }
+              if (testName === 'hooks fail') {
```
Comment (rochdev) [inline packages/datadog-plugin-cucumber/test/index.spec.js:198]: These conditions are starting to pile up in test framework tests. Any way to split that in multiple tests instead?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-plugin-cucumber/test/index.spec.js (hunk 2 above)
Comment (juan-fernandez) [reply packages/datadog-plugin-cucumber/test/index.spec.js:198]: yeah true. The setup made sense when all the tests were checking the same, but not anymore. I'll change it

Comment (juan-fernandez) [reply packages/datadog-plugin-cucumber/test/index.spec.js]: done for cucumber: https://github.com/DataDog/dd-trace-js/pull/1429/commits/0133537ab804cfb94418c222652f6759b40837ba
done for mocha: https://github.com/DataDog/dd-trace-js/pull/1429/commits/596fef9f7c8769e688624f55c4970bcdabc337b9

jest already had multiple tests

**Comment by juan-fernandez** [comment]
btw, `node-cucumber` will keep failing unless we merge https://github.com/DataDog/dd-trace-js/pull/1438 before

File: packages/datadog-plugin-jest/src/index.js [hunk 3]
```diff
@@ -139,6 +142,28 @@ function createHandleTestEvent (tracer, testEnvironmentMetadata, instrumenter) {
       testSpan.finish()
       return
     }
+    if (event.name === 'hook_failure') {
+      const testSpan = tracer.startSpan(
```
Comment (rochdev) [inline packages/datadog-plugin-jest/src/index.js:146]: Anyway way to wrap the hook to also get the timing information? What if the hook timed out or took 10 minutes before failing for example? The also applies to the other test frameworks.

Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js:146]: yeah that's a good point. Would it be OK to do it in a follow up PR? This is getting quite big already and it already solves the problem of identifying errors in the hooks.

Comment (juan-fernandez) [reply packages/datadog-plugin-jest/src/index.js:146]: there isn't one. For CI App, the trace will soon no longer be the entry point to a test, as we're seeing that most of the times (unit tests) we can only show a single span. Also, this trace will only have this span, so there'll be no problem with widths being too small in the UI.

In any case, I agree this needs to be solved. I just wonder if it'd be better to do it in a separate PR, so I don't have to keep rebasing /merge master on this PR.

**Review by rochdev — APPROVED** [review]

**Pull request merged**